nixos/modules/installer/tools/nixos-option/nixos-option.cc

```diff
@@ -0,0 +1,588 @@
+      out << "«not defined»";
+    } else {
+      out << describeError(e);
+    }
+  }
+}
+
+void printConfigValue(Context *ctx, Out &out, std::string const &path,
+                      std::variant<Value, Error> v) {
+  out << path << " = ";
+  printValue(ctx, out, std::move(v), path);
+  out << ";\n";
+}
+
+void printAll(Context *ctx, Out &out) {
+  mapOptions(
+      [ctx, &out](std::string const &option_path) {
+        mapConfigValuesInOption(
+            [ctx, &out](std::string const &config_path,
+                        std::variant<Value, Error> v) {
+              printConfigValue(ctx, out, config_path, v);
+            },
+            option_path, ctx);
+      },
+      ctx, ctx->options_root);
+}
+
+void printAttr(Context *ctx, Out &out, std::string const &path, Value *root) {
+  try {
+    printValue(ctx, out,
+               *findAlongAttrPath(*ctx->state, path, *ctx->autoArgs, *root),
+               path);
+  } catch (Error &e) {
+    out << describeError(e);
+  }
+}
+
+void printOption(Context *ctx, Out &out, std::string const &path,
+                 Value *option) {
+  out << "Value:\n";
+  printAttr(ctx, out, path, &ctx->config_root);
+
+  out << "\n\nDefault:\n";
+  printAttr(ctx, out, "default", option);
+
+  out << "\n\nType:\n";
+  printAttr(ctx, out, "type.description", option);
+
+  out << "\n\nExample:\n";
+  printAttr(ctx, out, "example", option);
+
+  out << "\n\nDescription:\n";
+  printAttr(ctx, out, "description", option);
+
+  out << "\n\nDeclared by:\n";
+  printAttr(ctx, out, "declarations", option);
+
+  out << "\n\nDefined by:\n";
+  printAttr(ctx, out, "files", option);
+  out << "\n";
+}
+
+void printListing(Out &out, Value *v) {
+  // Print this header on stderr rather than stdout because the old shell script
+  // implementation did.  I don't know why.
+  std::cerr << "This attribute set contains:\n";
+  for (const auto &a : v->attrs->lexicographicOrder()) {
+    std::string name = a->name;
+    if (!name.empty() && name[0] != '_') {
+      out << name << "\n";
+    }
+  }
+}
+
+// Carefully walk an option path, looking for sub-options when a path walks past
+// an option value.
+Value findAlongOptionPath(Context *ctx, std::string const &path) {
+  Strings tokens = parseAttrPath(path);
+  Value v = ctx->options_root;
+  for (auto i = tokens.begin(); i != tokens.end(); i++) {
+    bool last_attribute = std::next(i) == tokens.end();
+    auto const &attr = *i;
+    v = evaluateValue(ctx, &v);
+    if (attr.empty()) {
+      throw Error("empty attribute name in selection path '" + path + "'");
+    }
+    if (isOption(ctx, v) && !last_attribute) {
+      Value getSubOptions = evaluateValue(
+          ctx, findAlongAttrPath(*ctx->state, "type.getSubOptions",
+                                 *ctx->autoArgs, v));
+      if (getSubOptions.type != tLambda) {
+        throw Error("Option's type.getSubOptions isn't a function at '" + attr +
+                    "' in path '" + path + "'");
```

Comment:
Please use format strings instead of concatenation, e.g.
```
throw Error("option's 'type.getSubOptions' is not a function at '%s' in path '%s'", attr, path);
```